src/main/java/org/marketdesignresearch/mechlib/core/bid/bundle/BundleValueBid.java

```diff
@@ -0,0 +1,83 @@
+package org.marketdesignresearch.mechlib.core.bid.bundle;
+
+import java.math.BigDecimal;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.marketdesignresearch.mechlib.core.Bundle;
+import org.marketdesignresearch.mechlib.core.BundleEntry;
+import org.marketdesignresearch.mechlib.core.Good;
+import org.marketdesignresearch.mechlib.core.bid.Bid;
+import org.springframework.data.annotation.PersistenceConstructor;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+@ToString @EqualsAndHashCode
+public class BundleValueBid<T extends BundleValuePair> implements Bid{
+    public static BundleValueBid<BundleValuePair> singleBundleBid(BundleValuePair bundleBid) {
+        return new BundleValueBid<BundleValuePair>(ImmutableSet.of(bundleBid));
+    }
+
+    @Getter
+    private final Set<T> bundleBids;
+    
+    public BundleValueBid() {
+        this(new LinkedHashSet<>());
+    }
+    
+    @PersistenceConstructor
+    public BundleValueBid(Set<T> bundleBids) {
+    	// not more than one bid per bundle
+    	Preconditions.checkArgument(bundleBids.size() == bundleBids.stream().map(b -> b.getBundle()).collect(Collectors.toSet()).size());
+    	this.bundleBids = bundleBids;
+    }
+
+    public void addBundleBid(T bundleBid) {
+    	this.bundleBids.remove(this.getBidForBundle(bundleBid.getBundle()));
+        this.bundleBids.add(bundleBid);
+    }
+    
+    public T getBidForBundle(Bundle bundle) {
+    	return this.bundleBids.stream().filter(b -> b.getBundle().equals(bundle)).findAny().orElse(null);
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
This cast hurts a bit, but I remember it to be cumbersome to solve this properly with generics... Is there no other way? (this comment refers to the other cases of this suppressed warning as well)

Reply:
I agree this cast hurts. Finally I found a solution how to avoid this cast. Therefore I did some heavy changes to the bid.bundle package. This way makes it know very clean and straight forward to use I hope. See also my comments on the newest commit.

Reply:
👍